src/snowflake/snowpark/session.py

```diff
@@ -623,8 +620,19 @@ def __str__(self):
         )
 
     def _generate_new_action_id(self) -> int:
-        self._last_action_id += 1
-        return self._last_action_id
+        with self._lock:
+            self._last_action_id += 1
+            return self._last_action_id
+
+    @property
+    def _analyzer(self) -> Analyzer:
+        if not hasattr(self._thread_store, "analyzer"):
+            self._thread_store.analyzer = (
+                Analyzer(self)
+                if isinstance(self._conn, ServerConnection)
+                else MockAnalyzer(self)
+            )
+        return self._thread_store.analyzer
```

Comment:
same as above, do we need to protect this as well?

Reply:
same. this is a thread local object. this the how we create a per-thread instance of cursor and analyzer.